docs/adr/2023-11-07_029-streaming-persistence.md

```diff
@@ -0,0 +1,40 @@
+---
+slug: 29
+title: |
+  29. EventServer abstraction for event persistence
+authors: [@cardenaso11, @quantumplation]
+tags: []
+---
+
+## Status
+Draft
+
+## Context
+
+The Hydra node represents a significant engineering asset, providing layer 1 monitoring, peer to peer consensus, durable persistence, and an isomorphic Cardano ledger. Because of this, it is being eyed as a key building block not just in Hydra based applications, but other protocols as well.
+
+One remaining difficulty in integrating Hydra into a fully productionalized software stack is the persistence model.  Currently, the Hydra node achieves durability by writing a sequence of "StateChanged" events to disk. If a node is interrupted, upon restarting, it can replay just these events, ignoring their corresponding effects, to recover the same internal state it had when it was interrupted. However, downstream consumers don't have the same luxury.
+
+We propose generalizing the persistence mechanism to open the door to a plugin based approach to extending the Hydra node.
+
+# Decision
+
+We propose adding a "persistence combinator", which can combine one or more "persistenceIncremental" instances.
+
+When appending to the combinator, it will forward the append to each persistence mechanism.
+
+As the node starts up, as part of recovering the node state, it will also ensure "at least once" semantics for each persistence mechanism. It will maintain a notion of a "primary" instance, from which events are loaded, and "secondary instances", which are given the opportunity to re-observe each event.
```

Comment:
This is not entirely clear to me what you mean here, perhaps a diagram (sequence?) would help understanding how would that work? There's both the `append` and the `loadAll` part to consider.

Reply:
Imagine we have a setup that uses the CombinedPersistence mechanism to:
 - save the StateChanged to the log file
 - builds up a larger archive of transactions (say, 10,000 transactions at a time), and then saves them to S3
 - publishes a message for each StateChanged on kinesis

Now imagine we write the StateChanged event to disk, and crash when saving to S3.  On startup, we will load all events from StateChanged and replay them against the hydra state, recovering our place.  However:
 - the S3 destination wouldn't see the StateChanged event, and would lose progress on the most recent archive it was producing; downstream consumers would see a gap in the stream
 - The kinesis stream would also not see that last StateChanged event, and would also see a gap.

Perhaps, you might suggest, you solve this by reversing the order of these sinks: we save the event to each of the other sinks, and the persistence on disk last.  Now, when we recover, the state changed has never happened according to the node, but we've emitted it to S3 and kinesis, letting downstream actors react to state that the node then might not consider final.

So, if we build this combinator, it needs to *not* just feed the StateChanged events to the hydra state on startup (the outcomes of which get thrown away), but *also* to the other PersistenceIncremental instances. This gives an "at least once" delivery semantics for each PersistenceIncremental instance; This allows it to recover its partial state, and upgrade that to "exactly once" delivery semantics if the underlying medium support it (for example, checking if a file exists in S3, keeping track of it's own delivery cursor, etc.)